[ROCm] Enable group gemm through CK

jagadish-amd opened the pull request:
Fixes #161366
All the 4 types of dimension matrix are supported.
2d-2d, 2d-3d, 3d-3d, 3d-2d. The corresponding test cases in test_matmul_cuda are working
for both forward and backward pass.
The CK path is enabled for gfx942, gfx950. 
ToDo: Need to enable support on gfx90a since the ck kernel used in this commit produces gpu error, 
might require a different CK kernel config, based on the profiler result on gfx90a.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @naromero77amd

jagadish-amd committed:
Initial group gemm through CK code

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
Add working code, supports 2d*2d group gemm

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
Add dim logic to select 2d*2d case.

Also added comments.

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
Enable 2d*3d group gemm, forward pass.

python test/test_matmul_cuda.py -v -k test_grouped_gemm_2d_3d
Ran 24 tests in 5.566s
OK

python test/test_matmul_cuda.py -v -k test_grouped_gemm_2d_2d
Ran 24 tests in 5.537s

OK

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
Add 3d*3d case

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
Add 3d*2d case

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
Make sure input tensors are contiguous

All test cases are passing with forward and backward pass.

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
enable fast path / ck only for gfx942

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
Merge remote-tracking branch 'upstream/main' into group_gemm_ck

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
clean merge conflict

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
Unskip few tests in test_grouped_gemm_3d_2d

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
Verified group gemm ck support on gfx950

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd committed:
lint error: Update ck_group_gemm.h

jagadish-amd committed:
lint error: Update ck_group_gemm.hip

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

atalman commented:
Here are the errors:
```
ck_group_gemm.hip:129:13: error: missing field 'stride_Ds_' initializer [-Werror,-Wmissing-field-initializers]
  129 |             });
      |             ^

ck_group_gemm.hip:420:13: note: in instantiation of function template specialization 'at::hip::detail::launch_grouped_bgemm_ck_impl_dispatch<ck::tensor_layout::gemm::RowMajor, ck::tensor_layout::gemm::RowMajor, unsigned short>' requested here
  420 |             launch_grouped_bgemm_ck_impl_dispatch<ck::tensor_layout::gemm::RowMajor, ck::tensor_layout::gemm::RowMajor, CkTypes::BF16>(mat_a, mat_b, offs, out);
      |             ^
```

jagadish-amd committed:
Fix build error in ck_group_gemm.hip

error: missing field 'stride_Ds_' initializer [-Werror,-Wmissing-field-initializers]

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

jagadish-amd commented:
> Here are the errors:
> 
> ```
> ck_group_gemm.hip:129:13: error: missing field 'stride_Ds_' initializer [-Werror,-Wmissing-field-initializers]
>   129 |             });
>       |             ^
> 
> ck_group_gemm.hip:420:13: note: in instantiation of function template specialization 'at::hip::detail::launch_grouped_bgemm_ck_impl_dispatch<ck::tensor_layout::gemm::RowMajor, ck::tensor_layout::gemm::RowMajor, unsigned short>' requested here
>   420 |             launch_grouped_bgemm_ck_impl_dispatch<ck::tensor_layout::gemm::RowMajor, ck::tensor_layout::gemm::RowMajor, CkTypes::BF16>(mat_a, mat_b, offs, out);
>       |             ^
> ```

looking at the errors, it's due to
"missing field 'stride_Ds_' initializer"
Pushed a commit which initializes the field with empty vector.

jeffdaily commented:
@atalman how did the internal import go?

jagadish-amd committed:
Fix stride_Ds compiler error

Signed-off-by: Jagadish Krishnamoorthy <jagadish.krishnamoorthy@amd.com>

atalman reviewed (approved):
lgtm. Internal change is clear. Please fix lint:
[Lint / lintrunner-pyrefly-all / linux-job](https://hud.pytorch.org/pr/pytorch/pytorch/166334#54402142537) ([gh](https://github.com/pytorch/pytorch/actions/runs/19048255903/job/54402142537))
>>> Lint for torch/utils/tensorboard/writer.py:

jagadish-amd committed:
Merge branch 'main' of https://github.com/jagadish-amd/pytorch into group_gemm_ck

ScottTodd commented on aten/src/ATen/native/cuda/GroupedBlas.cpp:
```diff
@@ -666,12 +669,19 @@ std::optional<c10::ScalarType> out_dtype) {
     at::cuda::detail::bf16bf16_grouped_mm(mat_a, mat_b, offs, bias, out);
+#else
+    at::hip::detail::group_gemm_ck(mat_a, mat_b, offs, bias, out);
+#endif
```
This breaks builds on Windows, where CK is currently not enabled: https://github.com/ROCm/TheRock/issues/2054

```
2025-11-07T07:51:12.4977670Z [7076/7084] Linking CXX shared library bin\torch_hip.dll
2025-11-07T07:51:12.4977863Z FAILED: [code=4294967295] bin/torch_hip.dll lib/torch_hip.lib 
2025-11-07T07:51:12.4979510Z C:\Windows\system32\cmd.exe /C "cd . && C:\home\runner\_work\_tool\Python\3.12.10\x64\Lib\site-packages\cmake\data\bin\cmake.exe -E vs_link_dll --msvc-ver=1944 --intdir=caffe2\CMakeFiles\torch_hip.dir --rc="C:\Program Files (x86)\Windows Kits\10\bin\10.0.26100.0\x64\rc.exe" --mt="C:\Program Files (x86)\Windows Kits\10\bin\10.0.26100.0\x64\mt.exe" --manifests  -- C:\home\runner\_work\_tool\Python\3.12.10\x64\Lib\site-packages\_rocm_sdk_devel\lib\llvm\bin\lld-link.exe /nologo @CMakeFiles\torch_hip.rsp  /out:bin\torch_hip.dll /implib:lib\torch_hip.lib /pdb:bin\torch_hip.pdb /dll /version:0.0 /machine:x64 /ignore:4049 /ignore:4217 /ignore:4099 /INCREMENTAL:NO && cd ."
2025-11-07T07:51:12.4980262Z LINK: command "C:\home\runner\_work\_tool\Python\3.12.10\x64\Lib\site-packages\_rocm_sdk_devel\lib\llvm\bin\lld-link.exe /nologo @CMakeFiles\torch_hip.rsp /out:bin\torch_hip.dll /implib:lib\torch_hip.lib /pdb:bin\torch_hip.pdb /dll /version:0.0 /machine:x64 /ignore:4049 /ignore:4217 /ignore:4099 /INCREMENTAL:NO /MANIFEST:EMBED,ID=2" failed (exit code 1) with the following output:
2025-11-07T07:51:12.4980745Z lld-link: error: undefined symbol: void __cdecl at::hip::detail::group_gemm_ck(class at::Tensor const &, class at::Tensor const &, class std::optional<class at::Tensor> const &, class std::optional<class at::Tensor> const &, class at::Tensor &)
2025-11-07T07:51:12.4980868Z 
2025-11-07T07:51:12.4981526Z >>> referenced by caffe2\CMakeFiles\torch_hip.dir\__\aten\src\ATen\native\hip\GroupedBlas.cpp.obj:(class at::Tensor __cdecl at::native::_grouped_mm_cuda(class at::Tensor const &, class at::Tensor const &, class std::optional<class at::Tensor> const &, class std::optional<class at::Tensor> const &, class std::optional<enum c10::ScalarType>))
2025-11-07T07:51:12.4981649Z 
2025-11-07T07:51:12.4981729Z ninja: build stopped: subcommand failed.
```

Can this be conditioned on CK being enabled too?
https://github.com/pytorch/pytorch/blob/724cd32b0cab03c54fd1f288f7555cfe0544c531/CMakeLists.txt#L252-L253

Perhaps this code style? https://github.com/pytorch/pytorch/blob/724cd32b0cab03c54fd1f288f7555cfe0544c531/aten/src/ATen/cuda/CUDABlas.cpp#L871-L877

ScottTodd replied:
A fix-forward is in review at https://github.com/pytorch/pytorch/pull/167403. Can we revert first though? I'm not sure if I have permission to trigger the bot for that...

ScottTodd replied:
Confirmed that the fix-forward was successful and our nightly release builds are functional again. Latest status update at https://github.com/ROCm/TheRock/issues/2054#issuecomment-3528758141.
